### ISSUE TITLE
Hang when training LLama7B on TPU v5e-16

### ISSUE THREAD

**Issue by rjpower** [stanford-crfm/levanter#584]
I'm trying out Levanter with some v5e slices and I'm running into a few issues. The most important seems to be a hang that happens in the first few training steps. I don't think this is Levanter's fault, but curious if you've seen this before and have ideas for how to resolve it.

Evaluation finishes successfully, and then it appears that we get 1 or 2 training steps before hanging. (Sorry for the noise, I turned on TPU logging to see if the something was happening with XLA or the TPU runtime):

```
34.16.228.1: 2024-05-15T20:03:24 - 3 - levanter.eval - eval.py:120 - INFO :: eval loss: 11.185
34.16.228.1: 2024-05-15T20:03:24 - 3 - levanter.eval - eval.py:131 - INFO ::  loss: 11.185
...
34.16.214.62: I0515 20:03:26.569316  343449 flash_attention.cc:1908] successfully finished dry run
train:   0%|          | 2/100000 [00:46<644:59:38, 23.22s/it, loss=11.2]I0515 20:03:26.569367  443369 flash_attention.cc:1872] number of matches: 0
34.125.52.64: I0515 20:03:26.569430  443369 flash_attention.cc:1908] successfully finished dry run
34.16.214.62: I0515 20:03:26.571009  353969 memory_space_assignment_util.cc:241] Sliced prefetch options: go/debugonly
34.16.214.62: max_slices: 4
34.16.214.62: min_bytes: 1048576
34.16.214.62: all_slice_time_permutations_threshold: 4
34.125.52.64: I0515 20:03:26.571118  453899 memory_space_assignment_util.cc:241] Sliced prefetch options: go/debugproto
34.125.52.64: max_slices: 4
34.125.52.64: min_bytes: 1048576
34.125.52.64: all_slice_time_permutations_threshold: 4
34.16.149.229: I0515 20:03:26.570720  283245 flash_attention.cc:1908] successfully finished dry run
34.16.149.229: I0515 20:03:26.572364  293901 memory_space_assignment_util.cc:241] Sliced prefetch options: go/debugstr
34.16.149.229: max_slices: 4
34.16.149.229: min_bytes: 1048576
34.16.149.229: all_slice_time_permutations_threshold: 4
train:   0%|          | 2/100000 [00:46<644:55:25, 23.22s/it, loss=11.2]I0515 20:03:26.571827  415950 flash_attention.cc:1872] number of matches: 0
34.16.228.1: I0515 20:03:26.571900  415950 flash_attention.cc:1908] successfully finished dry run
```

Checking the Ray dashboard seems to imply that Levanter isn't doing anything wrong: all the jobs are waiting on PJRT:

```
Process 415584: python -m levanter.main.train_lm --config_path gs://wasabi-tpu-training/config/llama2_7b.yaml
Python v3.10.12 (/usr/bin/python3.10)

Thread 415584 (idle): "MainThread"
    _value (jax/_src/array.py:594)
    wrapper (jax/_src/profiler.py:335)
    __array__ (jax/_src/array.py:395)
    concrete_or_error (jax/_src/core.py:1494)
    _item (jax/_src/numpy/array_methods.py:78)
    train_step (levanter/trainer.py:372)
    training_steps (levanter/trainer.py:386)
    train (levanter/trainer.py:403)
    main (train_lm.py:198)
    wrapper_inner (levanter/config.py:84)
    <module> (train_lm.py:203)
    _run_code (runpy.py:86)
    _run_module_as_main (runpy.py:196)
```

The Levanter processes aren't CPU bound, and connecting with GDB doesn't show any suspicious threads that I can see. I'm using the Levanter at head from ~yesterday and the following jax/tpu version:

```
jax==0.4.28
jaxlib==0.4.28
jaxtyping==0.2.28
power@t1v-n-a0c5d808-w-0:~$ pip freeze | grep tpu
cloud-tpu-client==0.10
cloud-tpu-profiler==2.4.0
libtpu-nightly==0.1.dev20240508
```

I'm running with the stock llama2_7b.yaml configuration, just modified to use 1 shard of the validation dataset and adjust the cache directory. Command lines:

```
gcloud alpha compute tpus queued-resources create test-slice-16 --node-id test-slice-16 --zone=us-west4-a --accelerator-type=v5litepod-16 --runtime-version=v2-alpha-tpuv5-lite  --best-effort --project <project>

PDSH_SSH_ARGS="-A" pdsh -R ssh -w ^tpu-hosts 'GIT_SSH_COMMAND="ssh -oStrictHostKeyChecking=no" git clone git@github.com:rjpower/levanter'
PDSH_SSH_ARGS="-A" pdsh -R ssh -w ^tpu-hosts "cd levanter && pip install -e ."
PDSH_SSH_ARGS="-A" pdsh -R ssh -w ^tpu-hosts  "pip install jax[tpu] -f https://storage.googleapis.com/jax-releases/libtpu_releases.html"

# pip seems non-deterministic about dependency resolution, I got different versions from the default pip install
PDSH_SSH_ARGS="-A" pdsh -R ssh -w ^tpu-hosts 'pip install "ray==2.22.0"'
PDSH_SSH_ARGS="-A" pdsh -R ssh -w ^tpu-hosts 'export PATH=$PATH:$HOME/.local/bin/; export WANDB_API_KEY=...; export HF_TOKEN=..; export TPU_STDERR_LOG_LEVEL=0; export TPU_MIN_LOG_LEVEL=0; cd levanter && python -m levanter.main.train_lm --config_path gs://wasabi-tpu-training/config/llama2_7b.yaml'
```

I'm able to train gpt2_nano (though I get a warning about using the reference splash attention), so I suspect this is model specific. The "flash_attention.cc:1908] successfully finished dry run" seems suspicious but it could be a red-herring.

Please let me know if I can provide any more information!

**Comment by dlwh** [comment]
oof, thanks for the report. We also don't currently use jax 0.4.28 ourselves but 0.4.26, so there could be an issue there? Given that the hang is in _value I'm guessing it's an incompatibility with the latest JAX. You could also try `--model.use_flash_attention=false` to see if something went wrong with the TPU splash attention kernel. (We started using it recently and it could be there's an incompatibility with latest version of it.)

Can you just verify the commit hash? We fixed a couple of things recently and want to be sure, though the logging isn't consistent with those.

The two issues are a timeout in saving checkpoints inside tensorstore would cause a hang or crash, depending, and the other is I missed an error case when processing online. You're not seeing any messages about waiting for chunks?


I will try to repro soon!

Could you also try our setup scripts, just to see: https://github.com/stanford-crfm/levanter/blob/main/docs/Getting-Started-TPU-VM.md (They're kinda janky but they work pretty well) They don't currently use queued-resources but you could at least run infra/helpers/setup-tpu-vm.sh on each machine.

Good to know about pip non-determinism. I didn't realize that!  We should update to something more sane then.

For whatever reason, the TPU splash attention kernel only works with head dim 128, so we use my "plain JAX" version if that fails

**Comment by rjpower** [comment]
FWIW, I tried disabling flash attention in the model with:

```
model:
  use_flash_attention: False
  type: llama
```

But I still see references to flash attention in the logs, e.g.

```
 flash_attention.cc:1872] number of matches: 0
```

**Comment by dlwh** [comment]
I'm guessing disabling FA didn't fix in that case?

I've never seen the flash_attention.cc thing, but if I had to guess it's a compiler optimization pass in XLA?

**Comment by rjpower** [comment]
Thanks for the quick reply. Yeah, disabling flash didn't seem to help -- I tried editing the code just in case it was a configuration issue but got the same result.

You're right, the message is probably from an XLA pass. I wouldn't be surprised if XLA is trying to discover and automatically turn on Flash attention or something similar. I tried switching JAX back to 0.4.26, but I still saw it stalling. It's possible my environment is messed up on the VMs at this point though. I'm trying the infra script now to get a clean set of TPUs and see if that helps.

**Comment by dlwh** [comment]
(if you want to do a live debug ping my email which is GH handle @ stanford

**Comment by rjpower** [comment]
Sorry for the delay -- using infra setup worked! Thanks for the tip:

```
34.125.20.95: 2024-05-15T21:52:45 - 0 - levanter.eval - eval.py:120 - INFO :: eval loss: 7.331
34.125.183.145: 2024-05-15T21:52:45 - 3 - levanter.eval - eval.py:120 - INFO :: eval loss: 7.331
34.125.183.145: 2024-05-15T21:52:45 - 3 - levanter.eval - eval.py:131 - INFO ::  loss: 7.331
34.125.20.95: 2024-05-15T21:52:45 - 0 - levanter.eval - eval.py:131 - INFO ::  loss: 7.331
34.16.196.50: 2024-05-15T21:52:45 - 1 - levanter.eval - eval.py:120 - INFO :: eval loss: 7.331
34.125.242.121: 2024-05-15T21:52:45 - 2 - levanter.eval - eval.py:120 - INFO :: eval loss: 7.331
34.125.242.121: 2024-05-15T21:52:45 - 2 - levanter.eval - eval.py:131 - INFO ::  loss: 7.331
34.16.196.50: 2024-05-15T21:52:45 - 1 - levanter.eval - eval.py:131 - INFO ::  loss: 7.331
train:   0%|          | 201/100000 [13:39<99:38:53,  3.59s/it, loss=6.86]
train:   0%|          | 201/100000 [13:39<99:38:53,  3.59s/it, loss=6.86]
train:   0%|          | 201/100000 [13:39<99:38:55,  3.59s/it, loss=6.86]
```

Now I'm curious, I didn't realize the progress bar was only updating after 50 steps -- I'm going to retry with the original slice to see if I perhaps was just not waiting long enough... :/

**Comment by rjpower** [comment]
Hrm, seems to be working on the original slice with 0.4.26 and 0.4.28. Sorry for the noise!

It appears the progress bar is only being updated when we transition to eval: I originally adjusted steps_per_eval to reduce the amount of time per eval and probably exacerbated this issue. With this set to a high number, it looked like the system was hanging. I suspect because everything was working fine, but because the computation is so TPU bound, every time I tried to capture a stack trace, it would be waiting on PJRT.

It looks like the code should be updating the progress on each step, but I'm guessing there's some sort of buffering issue happening (with Ray?). When I added extra logging to the hook callback I see it being called in one big dump at the end once eval starts:

```
34.16.214.62: Running hooks:  5 1
34.16.228.1: Running hooks:  5 1
34.16.228.1: Running hooks:  5 2
34.16.149.229: Running hooks:  5 1
34.16.149.229: Running hooks:  5 2
34.16.149.229: Running hooks:  5 3
34.16.149.229: Running hooks:  5 4
```
